byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Chain.hs

```diff
@@ -202,32 +204,87 @@ sigGenChain
+    -- after a potential update in the protocol parameters (which is triggered
+    -- only at epoch boundaries). Otherwise the generators will use a state that
+    -- won't hold when the rules that correspond to these generators are
+    -- applied. For instance, the fees might change, which will render the
+    -- transaction as invalid.
+    --
+    let (us', _) = applySTSIndifferently @EPOCH $ TRC ( (sEpoch (Slot s) k, k)
+                                                      , us
+                                                      , nextSlot
+                                                      )
+
+        pps' = protocolParameters us'
+
+        upienv =
+          ( Slot s
+          , _dIStateDelegationMap ds
+          , k
+          , toNumberOfGenesisKeys $ Set.size ads
+          )
+
+        -- TODO: we might need to make the number of genesis keys a newtype, and
+        -- provide this function in the same module where this newtype is
+        -- defined.
+        toNumberOfGenesisKeys n
+          | fromIntegral (maxBound :: Word8) < n =
+              error $ "sigGenChain: too many genesis keys: " ++ show  n
+          | otherwise = fromIntegral n
+
+    aBlockVersion <-
+      Update.protocolVersionEndorsementGen upienv us'
+
     -- Here we do not want to shrink the issuer, since @Gen.element@ shrinks
     -- towards the first element of the list, which in this case won't provide
     -- us with better shrinks.
-    vkI         <- SigCntGen.issuer (pps, ds ^. dmsL, k) sgs
-    nextSlot    <- gNextSlot
+    vkI <- SigCntGen.issuer (pps', ds ^. dmsL, k) sgs
 
     delegationPayload <- case shouldGenDelegation of
       GenDelegation   ->
-        let dsEnv = DSEnv
-                    { _dSEnvAllowedDelegators = ads
-                    , _dSEnvEpoch = sEpoch nextSlot k
-                    , _dSEnvSlot = nextSlot
-                    , _dSEnvK = k
-                    }
-        in
-        dcertsGen dsEnv ds
+        -- In practice there won't be a delegation payload in every block, so we
+        -- make this payload sparse.
+        --
+        -- NOTE: We arbitrarily chose to generate delegation payload in 30% of
+        -- the cases. We could make this configurable.
+        Gen.frequency
+          [ (7, pure [])
+          , (3,
+              let dsEnv =
+                    DSEnv
+                      { _dSEnvAllowedDelegators = ads
+                      , _dSEnvEpoch = sEpoch nextSlot k
+                      , _dSEnvSlot = nextSlot
+                      , _dSEnvK = k
+                      }
+              in dcertsGen dsEnv ds
+            )
+          ]
       NoGenDelegation -> pure []
 
     utxoPayload <- case shouldGenUTxO of
-      GenUTxO   -> sigGen @UTXOWS Nothing utxoEnv utxo
+      GenUTxO   ->
+        let utxoEnv = UTxOEnv utxo0 pps' in
+          sigGen @UTXOWS Nothing utxoEnv utxo
       NoGenUTxO -> pure []
 
+    (anOptionalUpdateProposal, aListOfVotes) <-
```

Comment:
Yeah, I was being lazy here since we generate update payload only sporadically, but you're right, we should try to adhere to the principle of least surprise :)